### ISSUE TITLE
performance bug: ripgrep does poorly when given lots directory arguments

### ISSUE THREAD

**Issue by BurntSushi** [BurntSushi/ripgrep#44]
ripgrep is doing some non-trivial work for every file searched in positional parameters, and there's really no reason it should.

**Comment by BurntSushi** [comment]
In the `github.com/rust-lang/rust` repo, there is a pretty big time difference:

```
$ time rg E0401 | wc -l
7

real    0m0.072s
user    0m0.393s
sys     0m0.087s
$ time rg E0401 src/lib* | wc -l
6

real    0m0.202s
user    0m1.623s
sys     0m0.007s
```

If `-u` is passed to the second command, then the time difference goes away.

This means `ripgrep` is spending a lot of time reinitializing ignore state. There are some obvious things we shouldn't be redoing, like the globs and type filters from the command line, but those happen even with the `-u` flag, so they aren't the problem. The likely problem is re-traversing parent directories for `gitignore` files. Indeed, if `--no-ignore-parent` is given, then the performance difference goes away.

This means we need to keep our gitignores from parent directories around and try to reuse them.

**Comment by BurntSushi** [comment]
Fixed in #202.
